server/src/game/sessionManager.js

```diff
@@ -62,3 +79,38 @@ export async function getOrCreateActiveSession(user, questId) {
   }
   return createNewSession(user, questId)
 }
+
+export async function finishSession(session) {
+  session.status = 'finished'
+  await session.save()
+
+  const userProfile = await UserProfile.findOne({ user: session.user })
+  if (!userProfile) {
+    throw new Error(`UserProfile ${session.user} not found`)
+  }
+
+  const progress = userProfile.progress.get(session.quest)
+  if (!progress) {
+    throw new Error(`Progress ${session.quest} not found`)
+  }
+  if (!progress.completed) {
+    progress.finishedAt = new Date()
+    progress.completed = true
+  }
+  await userProfile.save()
+}
+
+export async function isQuestUnlocked(user, questId) {
+  const userProfile = await UserProfile.findOne({ user: user._id })
+  if (!userProfile) {
+    throw new Error(`UserProfile ${user._id} not found`)
+  }
+  const quest = await Quest.findById(questId)
+  if (!quest) {
+    throw new Error(`Quest ${questId} not found`)
+  }
+  return quest.requirements.every((requiredQuestId) => {
+    const progress = userProfile.progress.get(requiredQuestId)
+    return progress && progress.completed
+  })
```

Comment:
The progress variable here parsed the requirement already, what does this progress.completed contain?

Reply:
In `packages/models/userProfile.js`

```js
const progressSchema = new Schema({
  quest: {
    type: String,
    ref: 'Quest',
    required: true,
  },
  sessions: [
    {
      type: Schema.ObjectId,
      ref: 'Session',
      required: true,
    },
  ],
  completed: {
    type: Boolean,
    default: false,
    required: true,
  },
  startedAt: {
    type: Date,
    required: true,
  },
  finishedAt: Date,
})
```

A progress exists in UserProfile only implies that the quest is unlock and start being played, doesn't mean it is completed.

Reply:
Understood, thank you